pyro/distributions/planar.py

```diff
@@ -0,0 +1,123 @@
+from __future__ import absolute_import, division, print_function
+
+import math
+
+import torch
+import torch.nn as nn
+from torch.distributions.transforms import Transform
+from torch.distributions import constraints
+import torch.nn.functional as F
+
+from pyro.distributions.util import copy_docs_from
+
+
+@copy_docs_from(Transform)
+class PlanarFlow(Transform):
+    """
+    A 'planar' normalizing flow that uses the transformation
+
+        :math:`\\mathbf{y} = \\mathbf{x} + \\mathbf{u}\\tanh(\\mathbf{w}^T\\mathbf{z}+b)`
+
+    where :math:`\\mathbf{x}` are the inputs, :math:`\\mathbf{y}` are the outputs, and the learnable parameters
+    are :math:`b\\in\\mathbb{R}`, :math:`\\mathbf{u}\\in\\mathbb{R}^D`, :math:`\\mathbf{w}\\in\\mathbb{R}^D` for input
+    dimension :math:`D`. For this to be an invertible transformation, the condition
+    :math:`\\mathbf{w}^T\\mathbf{u}>-1` is enforced.
+
+    Together with `TransformedDistribution` this provides a way to create richer variational approximations.
+
+    Example usage:
+
+    >>> base_dist = dist.Normal(torch.zeros(10), torch.ones(10))
+    >>> plf = PlanarFlow(10)
+    >>> plf_module = pyro.module("my_plf", plf.module)
+    >>> plf_dist = dist.TransformedDistribution(base_dist, [plf])
+    >>> plf_dist.sample()  # doctest: +SKIP
+        tensor([-0.4071, -0.5030,  0.7924, -0.2366, -0.2387, -0.1417,  0.0868,
+                0.1389, -0.4629,  0.0986])
+
+    The inverse of this transform does not possess an analytical solution and is left unimplemented. However,
+    the inverse is cached when the forward operation is called during sampling, and so samples drawn using
+    planar flow can be scored.
+
+    :param input_dim: the dimension of the input (and output) variable.
+    :type autoregressive_nn: int
+
+    References:
+
+    Variational Inference with Normalizing Flows [arXiv:1505.05770]
+    Danilo Jimenez Rezende, Shakir Mohamed
+
+    """
+
+    codomain = constraints.real
+
+    def __init__(self, input_dim):
+        super(PlanarFlow, self).__init__()
+        self.input_dim = input_dim
+        self.module = nn.Module()
```

Comment:
Yeah that's a good question! It's so the parameters of the transform are picked up when you do `flow_dist.parameters()` on the `TransformedDistribution`, but I need to look more into how it works to understand if it's the best design to have a `self.module`

Reply:
@neerajprad @fritzo would it make sense to have some kind of `TransformedDistributionModuleMixIn`? 🤔

Reply:
I see. We could implement a mixin class that inherits off of `nn.module` and `TransformedDistribution`, but in that case `trans_dist.cuda()` will only end up sending the parameters of the transform to cuda() not the base distribution, so it doesn't seem like a clean separation. Another idea would be to have a `TransformModuleMixin` (inheriting from `Transform` and `nn.module`) which each of the flows can inherit from.

Reply:
Is this something that we should sort out with this pull request?

Reply:
When I make `PlanarFlow` inherit from `nn.Module` as well as `Transform` and remove the `.module` attribute, I get a `TypeError: unhashable type: 'PlanarFlow'` coming from `nn.module`. I'm going to look more into how this works so I can understand what's going on... Do you have any recommendations?

Reply:
Okay, so this behaviour is because `Transform` defines `.__eq__` but not `.__hash__`, which means that when you inherit from both `Transform` and `nn.Module` the `.__hash__` required for e.g. `nn.ModuleList` is not created automatically... I think this can be fixed easily

Reply:
@fritzo See the latest commit for my solution. Is this okay? Do you think I should do a pull request to PyTorch to get this changed in `Transform`? Otherwise, I'll fix `InverseAutoregressiveFlow` etc. for the same problem in another pull request